File: tests/wasm/test_epma.js
/**
 * WASM binding tests for EPMA indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly
 * and use the same reference values and tolerance.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
  loadTestData,
  assertArrayClose,
  isNaN,
  EXPECTED_OUTPUTS,
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
  
  const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
  const importPath = process.platform === 'win32'
    ? 'file:///' + wasmPath.replace(/\\/g, '/')
    : wasmPath;
  try {
    wasm = await import(importPath);
  } catch (err) {
    console.error('Failed to load WASM module. Run "wasm-pack build -- --features wasm --no-default-features" first');
    throw err;
  }
  testData = loadTestData();
});

test('EPMA accuracy (last 5 values)', () => {
  const close = new Float64Array(testData.close);
  const cfg = EXPECTED_OUTPUTS.epma;
  const out = wasm.epma_js(close, cfg.defaultParams.period, cfg.defaultParams.offset);
  assert.strictEqual(out.length, close.length);

  
  const last5 = out.slice(-5);
  assertArrayClose(last5, cfg.lastFive, 1e-1, 'EPMA last 5 values mismatch');
});

test('EPMA default candles length', () => {
  const close = new Float64Array(testData.close);
  const cfg = EXPECTED_OUTPUTS.epma;
  const out = wasm.epma_js(close, cfg.defaultParams.period, cfg.defaultParams.offset);
  assert.strictEqual(out.length, close.length);
});

test('EPMA invalid parameters', () => {
  const data = new Float64Array([10.0, 20.0, 30.0]);
  assert.throws(() => wasm.epma_js(data, 0, 4), /Invalid period|Invalid params|Invalid offset/i);
  assert.throws(() => wasm.epma_js(data, 10, 0), /Invalid period|Invalid params|Invalid offset/i);
});

test('EPMA empty and NaN input handling', () => {
  const empty = new Float64Array();
  assert.throws(() => wasm.epma_js(empty, 11, 4), /empty|length/i);

  const allNaN = new Float64Array(20);
  for (let i = 0; i < allNaN.length; i++) allNaN[i] = NaN;
  assert.throws(() => wasm.epma_js(allNaN, 11, 4), /All values are NaN/i);
});

test('EPMA warmup NaNs then valid values', () => {
  const close = new Float64Array(testData.close.slice(0, 200));
  const { defaultParams, warmupPeriod } = EXPECTED_OUTPUTS.epma;
  const out = wasm.epma_js(close, defaultParams.period, defaultParams.offset);
  
  for (let i = 0; i < Math.min(out.length, warmupPeriod); i++) {
    assert.ok(isNaN(out[i]), `Expected NaN during warmup at index ${i}`);
  }
  
  for (let i = warmupPeriod; i < out.length; i++) {
    assert.ok(!isNaN(out[i]), `Unexpected NaN after warmup at index ${i}`);
  }
});

test('EPMA batch API (default row check)', () => {
  const close = new Float64Array(testData.close);
  const cfg = EXPECTED_OUTPUTS.epma;
  const batch = wasm.epma_batch(close, {
    period_range: [cfg.defaultParams.period, cfg.defaultParams.period, 1],
    offset_range: [cfg.defaultParams.offset, cfg.defaultParams.offset, 1],
  });
  assert.strictEqual(batch.rows, 1);
  assert.strictEqual(batch.cols, close.length);
  assert.strictEqual(batch.values.length, batch.rows * batch.cols);
  assert.ok(Array.isArray(batch.combos) && batch.combos.length === 1);

  const row = batch.values.slice(batch.cols * 0, batch.cols * 1);
  const last5 = row.slice(-5);
  assertArrayClose(last5, cfg.lastFive, 1e-1, 'EPMA batch default row last 5 mismatch');
});

test('EPMA fast API (epma_into) matches safe API', () => {
  const close = new Float64Array(testData.close.slice(0, 256));
  const { period, offset } = EXPECTED_OUTPUTS.epma.defaultParams;

  const len = close.length;
  const inPtr = wasm.epma_alloc(len);
  const outPtr = wasm.epma_alloc(len);
  try {
    const mem = new Float64Array(wasm.__wasm.memory.buffer);
    const inOff = inPtr / 8;
    for (let i = 0; i < len; i++) mem[inOff + i] = close[i];

    wasm.epma_into(inPtr, outPtr, len, period, offset);

    const outOff = outPtr / 8;
    const fastOut = new Float64Array(len);
    for (let i = 0; i < len; i++) fastOut[i] = mem[outOff + i];

    const safeOut = wasm.epma_js(close, period, offset);
    assertArrayClose(fastOut, safeOut, 1e-10, 'EPMA fast API mismatch');
  } finally {
    wasm.epma_free(inPtr, len);
    wasm.epma_free(outPtr, len);
  }
});

File: tests/python/test_epma.py
"""
Python binding tests for EPMA indicator.
These tests mirror the Rust unit tests and use identical reference values
and tolerance (absolute 1e-1 on last-5 checks).
"""
import pytest
import numpy as np

try:
    import my_project as ti
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestEpma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_epma_accuracy(self, test_data):
        close = test_data['close']
        exp = EXPECTED_OUTPUTS['epma']

        out = ti.epma(close, exp['default_params']['period'], exp['default_params']['offset'])
        assert len(out) == len(close)

        
        assert_close(
            out[-5:],
            exp['last_5_values'],
            rtol=0.0,
            atol=1e-1,
            msg="EPMA last 5 values mismatch",
        )

        
        compare_with_rust('epma', out, 'close', exp['default_params'])

    def test_epma_invalid_params(self):
        data = np.array([10.0, 20.0, 30.0], dtype=np.float64)
        with pytest.raises(ValueError):
            ti.epma(data, 0, 4)
        with pytest.raises(ValueError):
            ti.epma(data, 10, 0)

    def test_epma_empty_and_nan_input(self):
        empty = np.array([], dtype=np.float64)
        with pytest.raises(ValueError):
            ti.epma(empty, 11, 4)

        all_nan = np.full(64, np.nan, dtype=np.float64)
        with pytest.raises(ValueError):
            ti.epma(all_nan, 11, 4)

    def test_epma_streaming(self, test_data):
        close = test_data['close']
        exp = EXPECTED_OUTPUTS['epma']
        p = exp['default_params']['period']
        o = exp['default_params']['offset']

        batch = ti.epma(close, p, o)
        stream = ti.EpmaStream(p, o)
        stream_vals = []
        for v in close:
            x = stream.update(float(v))
            stream_vals.append(np.nan if x is None else x)

        assert len(batch) == len(stream_vals)
        
        for i in range(len(batch)):
            if not (np.isnan(batch[i]) or np.isnan(stream_vals[i])):
                assert abs(batch[i] - stream_vals[i]) < 1e-9, \
                    f"Streaming mismatch at {i}: batch={batch[i]}, stream={stream_vals[i]}"

    def test_epma_batch(self, test_data):
        close = test_data['close']
        exp = EXPECTED_OUTPUTS['epma']
        p, o = exp['default_params']['period'], exp['default_params']['offset']

        
        batch = ti.epma_batch(close, (p, p, 1), (o, o, 1))
        assert 'values' in batch and 'periods' in batch and 'offsets' in batch
        assert batch['values'].shape == (1, len(close))
        assert list(batch['periods']) == [p]
        assert list(batch['offsets']) == [o]

        
        assert_close(
            batch['values'][0, -5:],
            exp['last_5_values'],
            rtol=0.0,
            atol=1e-1,
            msg='EPMA batch default row mismatch',
        )

        
        
        batch2 = ti.epma_batch(close[:200], (5, 11, 3), (1, 3, 2))  
        rows = len(batch2['periods'])
        assert batch2['values'].shape[0] == rows
        assert rows == len(batch2['offsets'])
        
        for i in range(rows):
            pi = int(batch2['periods'][i])
            oi = int(batch2['offsets'][i])
            single = ti.epma(close[:200], pi, oi)
            assert_close(batch2['values'][i], single, rtol=1e-10, atol=1e-12,
                         msg=f'EPMA batch row mismatch for period={pi}, offset={oi}')
